### PR TITLE
Issue675/dpl 215 {bug} handle isoseq samples in smrtlink

### LINKED ISSUE #675: DPL-215 - [BUG] Handle IsoSeq samples in SMRTLINK
**Describe the bug**
As a user (Craig) I would like to be able to generate sample sheets with IsoSeq samples in that can be uploaded to SMRTLINK without errors. A new version of SMRTLink (v10.2) prevents the IsoSeq sample work around we have. Email chain in comment on email with more details.

**RT Ticket Number**
If applicable

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to page '...'
2. Click on button '...'
3. See error

**Expected behaviour**

- [ ] “Sample is barcoded”: FALSE
- [ ] “Barcode Set”: Blank
- [ ] Sample level: Barcode name blank
- [ ] Sample level: Bio Sample name blank
- [ ] Well level: BioSample name present (concatenated?)

**Screenshots**

**Desktop (please complete the following information):**
* Windows or Mac?
* Browser Chrome, Firefox, Safari or other?
* Browser version (use 'About' to look up)?

**Additional context**

### PR DESCRIPTION
**Pull request by JamesGlover**
Closes #675

Changes proposed in this pull request:

*
*
* ...

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -2,7 +2,6 @@ require: rubocop-rails
 
 AllCops:
   NewCops: enable
-  TargetRubyVersion: 2.6
```
Comment (JamesGlover) [inline .rubocop.yml:5]: Not required when you have a ruby-version file, and can mean checks get out of sync.

File: Gemfile [hunk 2]
```diff
@@ -18,7 +18,6 @@ gem 'mysql2'
 gem 'puma', '~> 4.3' # Use Puma as the app server
 gem 'rack-cors' # Use Rack CORS for handling CORS, making cross-origin AJAX possible
 gem 'rails', '~> 6.0.4.1'
-gem 'rubocop-rails'
```
Comment (JamesGlover) [inline Gemfile:21]: Don't want to install in production

File: app/csv_generator/csv_generator.rb [hunk 3]
```diff
@@ -8,19 +8,18 @@ class CsvGenerator
   # configuration => Pipelines::Configuration::Item
   attr_accessor :run, :configuration
 
-  # return a CSV file
```
Comment (JamesGlover) [inline app/csv_generator/csv_generator.rb:11]: Incorrect documentation

File: app/csv_generator/csv_generator.rb [hunk 4]
```diff
@@ -8,19 +8,18 @@ class CsvGenerator
...
+  # return a CSV String
   # using run and configuration attributes
   # to generate headers and data
   def generate_sample_sheet
     CSV.generate do |csv|
       csv << csv_headers
-      sorted_wells = sort_wells(run.plate.wells)
```
Comment (JamesGlover) [inline app/csv_generator/csv_generator.rb:17]: Managed to simplify the sorting significantly.

File: app/csv_generator/csv_generator.rb [hunk 5]
```diff
@@ -8,19 +8,18 @@ class CsvGenerator
...
 
       sorted_wells.each do |well|
         # add well header row
         csv << csv_data(well: well, row_type: :well)
 
-        next unless well.all_libraries_tagged
```
Comment (JamesGlover) [inline app/csv_generator/csv_generator.rb:23]: In practice I don't think we'll see significant changes here, as we *shouldn't* see a mix of tagged and untagged libraries, but this feels like a safer failure state.

File: app/models/concerns/sample_sheet.rb [hunk 6]
```diff
@@ -1,74 +1,70 @@
 # frozen_string_literal: true
 
 # SampleSheet module
-# provides the group of methods
-# used to generate a pacbio run's sample sheet csv
-# TODO: This is included in a couple of places, but only some of the methods work in each.
```
Comment (JamesGlover) [inline app/models/concerns/sample_sheet.rb:6]: Did the todo

File: app/models/pacbio/well.rb [hunk 7]
```diff
@@ -25,13 +25,22 @@ class Well < ApplicationRecord
...
     validates :pre_extension_time, numericality: { only_integer: true }, allow_blank: true
 
+    def tag_set
+      tag_sets.first
+    end
+
+    def sample_sheet_behaviour
+      SampleSheetBehaviour.get(tag_set&.sample_sheet_behaviour || :untagged)
+    end
+
```
Comment (JamesGlover) [inline app/models/pacbio/well.rb:43]: This bit could potentially be dried out with the stuff in library

File: app/services/tag_set_service.rb [hunk 8]
```diff
@@ -16,6 +16,6 @@ def load_tag_set(tag_set_name)
 
     return if tag_set.nil?
 
-    loaded_tag_sets[tag_set_name] = tag_set.tags.map { |tag| [tag.oligo, tag.id] }.to_h
+    loaded_tag_sets[tag_set_name] = tag_set.tags.to_h { |tag| [tag.oligo, tag.id] }
```
Comment (JamesGlover) [inline app/services/tag_set_service.rb:19]: Rubocop

File: lib/tasks/create_tags.rake [hunk 9]
```diff
@@ -99,7 +99,11 @@ namespace :tags do
     task pacbio_isoseq: :environment do
       puts '-> Creating Pacbio IsoSeq tag set and tags'
       set = TagSet.pacbio_pipeline
+                  .create_with(sample_sheet_behaviour: :hidden)
                   .find_or_create_by!(name: 'IsoSeq_Primers_12_Barcodes_v1', uuid: 'd1bb7419-4343-286f-f6f7-7365fa2d1ee9')
+
+      set.update!(sample_sheet_behaviour: :hidden) unless set.hidden_sample_sheet_behaviour?
```
Comment (JamesGlover) [inline lib/tasks/create_tags.rake:105]: To actually set the correct flag.

File: lib/tasks/pacbio_data.rake [hunk 10]
```diff
@@ -15,54 +15,65 @@ namespace :pacbio_data do
     puts '-> Pacbio plates successfully created'
 
     puts '-> Creating pacbio libraries...'
```
Comment (JamesGlover) [inline lib/tasks/pacbio_data.rake:17]: This was more work than I expected. But the pools created by this task were all quite similar in composition, so this adds a bit of variety to make testing easier.

File: spec/csv_generator/csv_generator_spec.rb [hunk 11]
```diff
@@ -1,256 +1,226 @@
-require "rails_helper"
+require 'rails_helper'
 
 RSpec.describe CsvGenerator, type: :model do
-  after(:each) { File.delete('sample_sheet.csv') if File.exists?('sample_sheet.csv') }
```
Comment (JamesGlover) [inline spec/csv_generator/csv_generator_spec.rb:4]: Its not a file

File: spec/csv_generator/csv_generator_spec.rb [hunk 12]
```diff
@@ -1,256 +1,226 @@
...
-            sorted_wells.push(create(:pacbio_well, row: row, column: column))
-          end
-        end
-      plate = create(:pacbio_plate, wells: sorted_wells) 
-      run = create(:pacbio_run, plate: plate)
-      csv = ::CsvGenerator.new(run: run, configuration: Pipelines.pacbio.sample_sheet)
-
-      expect(csv.send(:sort_wells, sorted_wells.shuffle)).to eq(sorted_wells)
-    end
-  end
```
Comment (JamesGlover) [inline spec/csv_generator/csv_generator_spec.rb:254]: Stopped testing the private method, instead test the resulting public behaviour.

**Review by stevieing — APPROVED** [review]

**Pull request merged**